### ISSUE TITLE
PY_COLORS=1 causes invalid escape codes in --junitxml

### ISSUE THREAD

**Issue by quodlibetor** [pytest-dev/pytest#2560]
Issue: the combination of `PY_COLORS=1` and `--junitxml` causes some weird escaping in the generated xml. Specifically, it seems to partially escape the ANSI escape codes -- converting the `ESC` into `#`, but not doing anything else.

Running with just `--colors=yes` doesn't do this to the generated XML. (This issue really only shows up in my CI environment, where getting things fixed is potentially slow, and could "break" builds for people.) Since it seems like `PY_COLORS=1` should mean the same thing as `--color=yes` I think this is a bug.

This occurs for me with pytest versions 2.8 - 3.1. Pip list for my env with 3.1:

```
$ pip list
appdirs (1.4.3)
packaging (16.8)
pip (9.0.1)
py (1.4.34)
pyparsing (2.2.0)
pytest (3.1.2)
setuptools (34.3.3)
six (1.10.0)
wheel (0.29.0)
```

Example:

```
$ cat test.py
def test_example():
    assert False

############### the bug
$ PY_COLORS=1 pytest --junitxml=junit-env.xml test.py
=============================== test session starts ===============================
platform darwin -- Python 3.5.3, pytest-3.0.6, py-1.4.31, pluggy-0.4.0
rootdir: /Users/bwm/findable/virtualenvs/tmp-203699af9fd9e3ac, inifile:
plugins: cov-2.4.0
collected 1 items

test.py F

 generated xml file: /Users/bwm/findable/virtualenvs/tmp-203699af9fd9e3ac/junit-env.xml
==================================== FAILURES =====================================
__________________________________ test_example ___________________________________

    def test_example():
>       assert False
E       assert False

test.py:2: AssertionError
============================ 1 failed in 0.03 seconds =============================

$ cat junit-env.xml
<?xml version="1.0" encoding="utf-8"?><testsuite errors="0" failures="1" name="pytest" skips="0" tests="1" time="0.028"><testcase classname="test" file="test.py" line="0" name="test_example" time="0.0013718605041503906"><failure message="assert False">#x1B[1m    def test_example():#x1B[0m
#x1B[1m&gt;       assert False#x1B[0m
#x1B[1m#x1B[31mE       assert False#x1B[0m

#x1B[1m#x1B[31mtest.py#x1B[0m:2: AssertionError</failure></testcase></testsuite>%
```

```
############## desired behavior
$ pytest --color=yes --junitxml=junit-flag.xml test.py
=============================== test session starts ===============================
platform darwin -- Python 3.5.3, pytest-3.0.6, py-1.4.31, pluggy-0.4.0
rootdir: /Users/bwm/findable/virtualenvs/tmp-203699af9fd9e3ac, inifile:
plugins: cov-2.4.0
collected 1 items

test.py F

 generated xml file: /Users/bwm/findable/virtualenvs/tmp-203699af9fd9e3ac/junit-flag.xml
==================================== FAILURES =====================================
__________________________________ test_example ___________________________________

    def test_example():
>       assert False
E       assert False

test.py:2: AssertionError
============================ 1 failed in 0.03 seconds =============================

$ cat junit-flag.xml
<?xml version="1.0" encoding="utf-8"?><testsuite errors="0" failures="1" name="pytest" skips="0" tests="1" time="0.027"><testcase classname="test" file="test.py" line="0" name="test_example" time="0.0012619495391845703"><failure message="assert False">def test_example():
&gt;       assert False
E       assert False

test.py:2: AssertionError</failure></testcase></testsuite>%
```

**Comment by nicoddemus** [comment]
Indeed looks like a bug, thanks!

**Comment by RonnyPfannschmidt** [comment]
this is a bug in pylib

**Comment by eric-wieser** [comment]
Fix looks straightforward, output should be `&#x1B;` not `#x1B`.

**Comment by DerrickRice** [comment]
Agreed w/ @eric-wieser , and strong disagree with @RonnyPfannschmidt  - looks like this is very much pytest code. The following lines escape a unicode into `#x0000` rather than `&#x0000`. Note the missing `&`.

https://github.com/pytest-dev/pytest/blob/master/src/_pytest/junitxml.py#L57-L65

**Comment by DerrickRice** [comment]
In combination:
- junit xml capture of logs
- logs getting colored based on `--color` and no way to disable this without also disabling console color (which I'd like to keep)
- improper escaping

Leads to an XML file that's a mess.

**Comment by eric-wieser** [comment]
My comment is wrong, `&x#0000;` is not legal xml.

**Comment by DerrickRice** [comment]
https://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references 

Can you elaborate? Looks like it's proper XML and HTML.

**Comment by DerrickRice** [comment]
Though it should also have the semicolon, which it currently does not. So it's missing both the ampersand and the semicolon. (The semicolon that we see is literal from "B;"

**Comment by eric-wieser** [comment]
See the closed PR above that references this issue.

**Comment by DerrickRice** [comment]
Oh. **specifically `&#x0;` is invalid. It's an accurate escaping of the null character, and the null character is disallowed.

`#x1B` should still be `&x1;B`, though.

**Comment by eric-wieser** [comment]
> and the null character is disallowed.

Not just the null character. XML 1.0 forbids a wide range of characters, escaped or otherwise.

**Comment by eric-wieser** [comment]
https://github.com/pytest-dev/pytest/pull/7107#issuecomment-617653467

**Comment by DerrickRice** [comment]
Ah, I see. unicode `1B` isn't in the valid range, and it appears (based on the src I quoted before) that this is only applied for characters in the invalid range.

To summarize the issue, then:

Pytest junitxml has to handle all possible characters, even those which may not be valid XML characters (even when escaped) and possibly even those which aren't valid unicode. It's current choice is to rewrite those characters as `x#__` where `__` is the hexidecimal character value. This looks suspiciously similar to the XML escaping `&#x__;` but it is not intended to be that.

A "fix" here might be to allow junitxml to configure it's escaping mechanism. For example, python provides 'ignore', 'replace', 'xmlcharrefreplace', 'backslashreplace' and any other name registered via codecs.register_error(). Obviously some of those don't help, but providing similar options would be valuable.

A more context-aware fix might be to remove color escapes entirely. No other solution is going to remove the non-unicode characters that follow the unicode [CSI](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Functions-using-CSI-_-ordered-by-the-final-character_s_). pytest already knows what those are. Pytest put them there, with the ColorizedFormatter.

Other CSI-style escapes can show up in stdout and stderr captures from downstream tools, as well. It may be helpful to be able to remove those.

This issue may be miscategorized. It seems like a choice rather than a bug.
